solr/core/src/java/org/apache/solr/pkg/PackageAPI.java

```diff
@@ -211,7 +211,7 @@ public boolean equals(Object obj) {
 
     @Override
     public int hashCode() {
-      throw new UnsupportedOperationException("TODO unimplemented");
+      return Objects.hash(version, manifestSHA512);
```

Comment:
`equals()` above does not compare `manifestSHA512`.
Unless `manifestSHA512` are equal if `version` are (in which case no need to add it here?), we could have two objets that are `equals()` yet have different `hashCode()`, which would be against the spec (see `java.lang.Object.hashCode()` javadoc).

Reply:
WDYT about just version here? If we include files, computing the hashcode seems like it could get unreasonably expensive. It's not clear to me how long that list could be.

Comparing two different packages that have the same version would break equals too so this seems safe.

Reply:
I believe that when `hashCode()` is called (when using a collection) it is followed by a call to `equals()` if there's an element in that slot. More efficient (better spreading) `hashCode()` would save calls to `equals()`. Therefore, I'm not sure saving on execution time on `hashCode()` makes sense, as we will have to pay it in the cost of computing more `equals()`.
It might pay in cases where the collections will contain very few elements, in which case `hashCode()` is not really important, but in these cases its performance is likely not an issue either.

The approach I would take here is do the right thing in implementing `hashCode()`, then if we have performance issues with some objects, understand what those are and solve them (_insert Donald Knuth quote here_).

Therefore I would base `hashCode()` on version and files like `equals()`.

Reply:
What I wrote above might not hold. Given that `equals()` does:

`        return Objects.equals(this.version, that.version)
            && Objects.equals(this.files, that.files);
`
it is not computing the equality of `files` when `version` is different, so indeed it might be better to not use `files` in `hashCode()`.

Reply:
Thanks,